File: services/bootstrap.py
"""Unified bootstrap — one load sequence for the iCUE interface."""

from dataclasses import dataclass
from typing import Callable, Optional

from core.engine import MacroManager
from services.engine_proxy import EngineProxy
from services.node_bridge import NodeBridge
from services.profile_manager import ProfileManager
from services.sidecar_api import SidecarAPI
from utils.debug import log

ProgressCallback = Callable[[int, str], None]


@dataclass
class BootContext:
    manager: MacroManager
    proxy: EngineProxy
    profiles: ProfileManager
    sidecar: SidecarAPI
    node: NodeBridge


def _report(cb: Optional[ProgressCallback], percent: int, message: str):
    if cb:
        cb(percent, message)


def bootstrap(progress: Optional[ProgressCallback] = None) -> BootContext:
    """Initialize engine, profiles, sidecar and Node.js in one sequence."""
    log("BOOT", "démarrage bootstrap")

    _report(progress, 5, "Initialisation du moteur Win32…")
    manager = MacroManager()
    log("BOOT", "moteur OK")

    _report(progress, 25, "Connexion des macros souris / clavier…")
    proxy = EngineProxy(manager)
    log("BOOT", "proxy OK")

    _report(progress, 45, "Chargement du profil…")
    profiles = ProfileManager()
    profiles.load("default")
    profiles.apply_to_engine(manager)
    log("BOOT", "profil OK")

    _report(progress, 60, "Démarrage API Sidecar (port 17840)…")
    sidecar = SidecarAPI(proxy, profiles=profiles)
    sidecar.start()
    log("BOOT", f"sidecar online={sidecar.online}")

    _report(progress, 75, "Démarrage Node.js (port 5173)…")
    node = NodeBridge()
    node.start()
    log("BOOT", f"node online={node.online}")

    _report(progress, 90, "Préparation interface iCUE…")

    return BootContext(
        manager=manager,
        proxy=proxy,
        profiles=profiles,
        sidecar=sidecar,
        node=node,
    )

File: cpp/src/app/bootstrap.cpp
#include "app/bootstrap.hpp"

#include "config/paths.hpp"
#include "core/engine.hpp"
#include "core/rgb_engine.hpp"
#include "services/engine_proxy.hpp"
#include "services/profile_manager.hpp"
#include "services/sidecar_api.hpp"

namespace gx {

BootContext bootstrap(const std::wstring& root) {
    if (!root.empty()) set_project_root(root);

    BootContext ctx;
    ctx.manager = std::make_unique<MacroManager>();
    ctx.rgb = std::make_unique<RgbEngine>();
    ctx.proxy = std::make_unique<EngineProxy>(*ctx.manager);
    ctx.profiles = std::make_unique<ProfileManager>();
    ctx.profiles->load("default");
    ctx.profiles->apply_to_engine(*ctx.manager);
    ctx.profiles->apply_to_rgb(*ctx.rgb);
    ctx.sidecar = std::make_unique<SidecarAPI>(*ctx.proxy, ctx.profiles.get());
    ctx.sidecar->start();
    return ctx;
}

}  // namespace gx
